pkg/ucp/store/inmemory/client.go

```diff
@@ -0,0 +1,262 @@
+/*
+Copyright 2023 The Radius Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package inmemory
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/radius-project/radius/pkg/ucp/resources"
+	"github.com/radius-project/radius/pkg/ucp/store"
+	"github.com/radius-project/radius/pkg/ucp/store/storeutil"
+	"github.com/radius-project/radius/pkg/ucp/util/etag"
+	"golang.org/x/exp/maps"
+)
+
+var _ store.StorageClient = (*Client)(nil)
+
+// Client is an in-memory implementation of store.StorageClient.
+type Client struct {
+	// mutex is used to synchronize access to the resources map.
+	mutex sync.Mutex
+
+	// resources is a map of resource IDs to their corresponding entries.
+	//
+	// The Get/Save/Delete methods will use the resource ID directly since they only access
+	// a single entry at a time.
+	//
+	// The Query method will iterate over all entries in the map to find the matching ones.
+	resources map[string]entry
+}
+
+// entry stores the commonly-used fields (extracted from the resource ID) for comparison in queries.
+// This is provided for ease of debugging.
+//
+// Here's an example:
+//
+//	resource ID: "/planes/radius/local/resourceGroups/my-rg/providers/Applications.Test/testType1/my-resource/testType2/my-child-resource"
+//
+// The entry would be:
+//
+//	rootScope: "/planes/radius/local/resourceGroups/my-rg"
+//	resourceType: "Applications.Test/testType1/testType2"
+//	routingScope: "Applications.Test/testType1/my-resource/testType2/my-child-resource"
+//
+// All fields are compared case-insensitively.
+type entry struct {
+	obj store.Object
+
+	// rootScope is the root scope of the resource ID.
+	rootScope string
+
+	// resourceType is the resource type of the resource ID.
+	resourceType string
+
+	// routingScope is the routing scope of the resource ID.
+	routingScope string
+}
+
+// NewClient creates a new in-memory store client.
+func NewClient() *Client {
+	return &Client{
+		mutex:     sync.Mutex{},
+		resources: map[string]entry{},
+	}
+}
+
+// Get implements store.StorageClient.
```

Comment:
Do the docs need an update?

Reply:
Is there something interesting you want to be documented here? I don't think it's useful to copy-paste things.

Reply:
We talked offline and agreed not to duplicate the StorageClient documentation. I also saw that `StorageClient` docs were missing, so I wrote some.